distributed/tests/test_client.py

```diff
@@ -3592,65 +3592,56 @@ async def test_scatter_raises_if_no_workers(c, s):
 async def test_reconnect():
     port = open_port()
 
-    async def hard_stop(s):
-        for pc in s.periodic_callbacks.values():
-            pc.stop()
-
-        s.stop_services()
-        for comm in list(s.stream_comms.values()):
-            comm.abort()
-        for comm in list(s.client_comms.values()):
-            comm.abort()
-
-        await s.rpc.close()
-        s.stop()
-        await Server.close(s)
-
-    async with Scheduler(port=port) as s:
-        async with Client(f"127.0.0.1:{port}", asynchronous=True) as c:
-            async with Worker(f"127.0.0.1:{port}") as w:
-                await c.wait_for_workers(1, timeout=10)
-                x = c.submit(inc, 1)
-                assert (await x) == 2
-                await hard_stop(s)
-
-                start = time()
-                while c.status != "connecting":
-                    assert time() < start + 10
-                    await asyncio.sleep(0.01)
-
-                assert x.status == "cancelled"
-                with pytest.raises(CancelledError):
-                    await x
-
-                async with Scheduler(port=port) as s2:
-                    start = time()
-                    while c.status != "running":
-                        await asyncio.sleep(0.1)
-                        assert time() < start + 10
-
-                    await w.finished()
-                    async with Worker(f"127.0.0.1:{port}"):
-                        start = time()
-                        while len(await c.nthreads()) != 1:
-                            await asyncio.sleep(0.05)
-                            assert time() < start + 10
-
-                        x = c.submit(inc, 1)
-                        assert (await x) == 2
-                        await hard_stop(s2)
-
-                start = time()
-                while True:
-                    assert time() < start + 10
-                    try:
-                        await x
-                        assert False
-                    except CommClosedError:
-                        continue
-                    except CancelledError:
-                        break
-            await c._close(fast=True)
+    stack = ExitStack()
+    proc = popen(["dask-scheduler", "--no-dashboard", f"--port={port}"])
+    stack.enter_context(proc)
+    async with Client(f"127.0.0.1:{port}", asynchronous=True) as c, Worker(
+        f"127.0.0.1:{port}"
+    ) as w:
+        await c.wait_for_workers(1, timeout=10)
+        x = c.submit(inc, 1)
+        assert (await x) == 2
+        stack.close()
```

Comment:
```suggestion
    async with Client(f"127.0.0.1:{port}", asynchronous=True) as c, Worker(
        f"127.0.0.1:{port}"
    ) as w:
        with popen(["dask-scheduler", "--no-dashboard", f"--port={port}"]):
            await c.wait_for_workers(1, timeout=10)
            x = c.submit(inc, 1)
            assert (await x) == 2
```
nit: I think you could write it without the ExitStack, since the client and workers will wait for the scheduler to come up. Not necessary though, just a little easier to read to me.

Reply:
you are right, they are waiting. they are only running and enter the body of the ctx manager once they are started. your suggestion would deadlock and timeout